### ISSUE TITLE
[bug]: .SPM duplicate files while processing

### ISSUE THREAD

**Issue by phooda1** [AFM-SPM/TopoStats#1391]
### Checklist

- [x] Run using the `main` branch to check if the issue persists. Instructions on how to do this can be found [here](https://afm-spm.github.io/TopoStats/main/installation.html#installing-from-github).
- [x] Find the offending file in the output. If processing halts, re-run analysis with `topostats --core 1 process`.
- [x] Describe the bug.
- [x] Include the configuration file.
- [x] Copy of the log-file from running with `topostats --log-level debug <command>`.
- [x] The exact command that failed. This is what you typed at the command line, including any options.
- [x] TopoStats version, this is reported by `topostats --version`
- [x] Operating System and Python Version

### Describe the bug

In the processed output folder, the folder for each image is being stored as .spm, hence after running the first time, if I re process my images, the number of .spm files that are detected doubles - as the folder .spm are also being recognised as spm files. Not sure if this is causing issues in processing but it does make the processing a bit longer.

### Copy of the log-file from running with `topostats --log-level debug <command>`

`[Tue, 04 Aug 2026 20:49:10] [INFO    ] [topostats] [__init__.py] [41] TopoStats version : 2.3.3.dev332+gacf1317f7
[Tue, 04 Aug 2026 20:49:10] [INFO    ] [topostats] [__init__.py] [42] Commit            : acf1317f7
[Tue, 04 Aug 2026 20:49:10] [INFO    ] [topostats] [validation.py] [32] The YAML configuration file is valid.
[Tue, 04 Aug 2026 20:49:10] [INFO    ] [topostats] [validation.py] [32] The YAML plotting configuration file is valid.
[Tue, 04 Aug 2026 20:49:10] [INFO    ] [topostats] [processing.py] [1581] Configuration run options are consistent, processing can proceed.
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [run_modules.py] [97] Configuration file loaded from      : RNA_config.yaml
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [run_modules.py] [98] Scanning for images in              : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [run_modules.py] [99] Output directory                    : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [run_modules.py] [100] Looking for images with extension   : .spm
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [run_modules.py] [101] Images with extension .spm in /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4 : 4
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [run_modules.py] [106] Thresholding method (Filtering)     : std_dev
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [run_modules.py] [107] Thresholding method (Grains)        : absolute
[Tue, 04 Aug 2026 20:49:20] [INFO    ] [topostats] [io.py] [790] Extracting image from /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/eGFPmRNAinTEAA_8ng_APSwaterincubateimage.0_00022.spm
[Tue, 04 Aug 2026 20:49:21] [INFO    ] [topostats] [io.py] [790] Extracting image from /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/eGFPmRNAinTEAA_8ng_APSwaterincubateimage.0_00024.spm
[Tue, 04 Aug 2026 20:51:14] [INFO    ] [topostats] [run_modules.py] [240] [eGFPmRNAinTEAA_8ng_APSwaterincubateimage.0_00024.spm] Processing completed.
[Tue, 04 Aug 2026 21:01:13] [INFO    ] [topostats] [run_modules.py] [240] [eGFPmRNAinTEAA_8ng_APSwaterincubateimage.0_00022.spm] Processing completed.
[Tue, 04 Aug 2026 21:01:14] [INFO    ] [topostats] [run_modules.py] [300] Saved grain stats to : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/grain_statistics.csv.
[Tue, 04 Aug 2026 21:01:15] [INFO    ] [topostats] [run_modules.py] [321] Saved matched branch stats to : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/matched_branch_statistics.csv.
[Tue, 04 Aug 2026 21:01:15] [INFO    ] [topostats] [run_modules.py] [339] Saved disordered tracing stats to : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/branch_statistics.csv.
[Tue, 04 Aug 2026 21:01:16] [INFO    ] [topostats] [run_modules.py] [353] Saved molecule stats to : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/molecule_statistics.csv.
[Tue, 04 Aug 2026 21:01:18] [INFO    ] [topostats] [validation.py] [32] The YAML summarisation config is valid.
[Tue, 04 Aug 2026 21:01:18] [INFO    ] [topostats] [run_modules.py] [390] [plotting] Default variable to labels mapping loaded.
[Tue, 04 Aug 2026 21:01:18] [INFO    ] [topostats] [run_modules.py] [399] Summary plots and statistics will be saved to : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/summary_distributions
[Tue, 04 Aug 2026 21:01:35] [INFO    ] [topostats] [processing.py] [1669] 

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  TopoStats Version           : 2.3.3
  TopoStats Commit            : acf1317f7
  Base Directory              : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4
  File Extension              : .spm
  Files Found                 : 4
  Successfully Processed^1    : 2 (50.0%)
  All statistics              : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/grain_statistics.csv
  Distribution Plots          : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/summary_distributions

  Configuration               : /Volumes/shared/pyne_group/Shared/AFM_Data/RNA_Structures/Preksha/RNA Notebook/v4/config.yaml

  Email                       : topostats@sheffield.ac.uk
  Documentation               : https://afm-spm.github.io/topostats/
  Source Code                 : https://github.com/AFM-SPM/TopoStats/
  Bug Reports/Feature Request : https://github.com/AFM-SPM/TopoStats/issues/new/choose
  Citation File Format        : https://github.com/AFM-SPM/TopoStats/blob/main/CITATION.cff

  ^1 Successful processing of an image is detection of grains and calculation of at least
     grain statistics. If these have been disabled the percentage will be 0.

  If you encounter bugs/issues or have feature requests please report them at the above URL
  or email us.

  If you have found TopoStats useful please consider citing it. A Citation File Format is
  linked above and available from the Source Code page.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

### Include the configuration file

[RNA_config.yaml](https://github.com/user-attachments/files/30748959/RNA_config.yaml)

### TopoStats Version

2.3.2rc2

### Python Version

3.11

### Operating System

MacOS M1/M2 (post-2021)

### Python Packages

absl-py==2.3.1
AFMReader==0.0.6
annotated-types==0.7.0
appnope @ file:///home/conda/feedstock_root/build_artifacts/appnope_1733332318622/work
art==6.5
asttokens @ file:///home/conda/feedstock_root/build_artifacts/asttokens_1763409923949/work
astunparse==1.6.3
biopython==1.86
certifi==2025.10.5
charset-normalizer==3.4.4
cheap_repr==0.5.2
comm @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_comm_1753453984/work
contourpy==1.3.3
cycler==0.12.1
debugpy @ file:///opt/miniconda3/conda-bld/debugpy_1762421636036/work
decorator==5.2.1
docstring_parser==0.17.0
et_xmlfile==2.0.0
executing @ file:///home/conda/feedstock_root/build_artifacts/executing_1756729339227/work
flatbuffers==25.9.23
fonttools==4.60.1
gast==0.6.0
google-pasta==0.2.0
grpcio==1.76.0
h5py==3.15.1
idna==3.11
igor2==0.5.12
ImageIO==2.37.2
ipykernel @ file:///Users/runner/miniforge3/conda-bld/bld/rattler-build_ipykernel_1770566135/work
ipython==8.37.0
ipython_pygments_lexers @ file:///home/conda/feedstock_root/build_artifacts/ipython_pygments_lexers_1737123620466/work
jedi @ file:///home/conda/feedstock_root/build_artifacts/jedi_1733300866624/work
joblib==1.5.2
jupyter_client @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_jupyter_client_1767954655/work
jupyter_core @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_jupyter_core_1760643864/work
keras==3.12.0
kiwisolver==1.4.9
lazy_loader==0.4
libclang==18.1.1
llvmlite==0.45.1
loguru==0.7.3
magicgui==0.10.1
Markdown==3.10
markdown-it-py==4.0.0
MarkupSafe==3.0.3
matplotlib==3.10.7
matplotlib-inline @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-inline_1778264321245/work
mdurl==0.1.2
ml_dtypes==0.5.3
namex==0.1.0
nest_asyncio @ file:///home/conda/feedstock_root/build_artifacts/nest-asyncio_1733325553580/work
networkx==3.5
numba==0.62.1
numpy==2.0.2
numpyencoder==0.3.2
openpyxl==3.1.5
opt_einsum==3.4.0
optree==0.17.0
packaging @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_packaging_1777103621/work
pandas==2.3.3
parso==0.8.5
patsy==1.0.2
pexpect @ file:///home/conda/feedstock_root/build_artifacts/pexpect_1733301927746/work
pillow==11.1.0
platformdirs @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_platformdirs_1775741140/work
prompt_toolkit @ file:///home/conda/feedstock_root/build_artifacts/prompt-toolkit_1756321756983/work
protobuf==6.33.0
psutil==5.9.8
psygnal==0.15.0
ptyprocess @ file:///home/conda/feedstock_root/build_artifacts/ptyprocess_1733302279685/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl#sha256=92c32ff62b5fd8cf325bec5ab90d7be3d2a8ca8c8a3813ff487a8d2002630d1f
pure_eval @ file:///home/conda/feedstock_root/build_artifacts/pure_eval_1733569405015/work
pyconify==0.2.1
pydantic==2.13.4
pydantic_core==2.46.4
Pygments @ file:///home/conda/feedstock_root/build_artifacts/pygments_1774796700044/work
pyparsing==3.2.5
pyspm==0.6.3
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_python-dateutil_1751104122/work
pytz==2025.2
PyYAML==6.0.3
pyzmq @ file:///opt/miniconda3/conda-bld/pyzmq_1762375616040/work
QtPy==2.4.3
requests==2.32.5
rich==14.2.0
ruamel.yaml==0.18.16
ruamel.yaml.clib==0.2.14
schema==0.7.8
scikit-image==0.25.2
scikit-learn==1.7.2
scikit-posthocs==0.12.0
scipy==1.16.3
seaborn==0.13.2
six @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_six_1753199211/work
skan==0.13.0
snoop==0.6.0
stack_data @ file:///home/conda/feedstock_root/build_artifacts/stack_data_1733569443808/work
statsmodels==0.14.6
superqt==0.7.6
tensorboard==2.20.0
tensorboard-data-server==0.7.2
tensorflow==2.20.0
termcolor==3.2.0
threadpoolctl==3.6.0
tifffile==2025.10.16
toolz==1.1.0
topoly==1.1.0
-e git+https://github.com/AFM-SPM/TopoStats.git@acf1317f7597026c651487b68c9fa4ea72bdf34a#egg=topostats
tornado @ file:///Users/ec2-user/croot/tornado_1774303634539/work
tqdm==4.67.1
traitlets==5.14.3
typing-inspection==0.4.2
typing_extensions @ file:///home/conda/feedstock_root/build_artifacts/bld/rattler-build_typing_extensions_1756220668/work
tzdata==2025.2
urllib3==2.5.0
wcwidth @ file:///home/conda/feedstock_root/build_artifacts/wcwidth_1777744409267/work
Werkzeug==3.1.3
wrapt==2.0.1

**Comment by ns-rse** [comment]
> In the processed output folder, the folder for each image is being stored as .spm, hence after running the first time,
> if I re process my images, the number of .spm files that are detected doubles - as the folder .spm are also being
> recognised as spm files. Not sure if this is causing issues in processing but it does make the processing a bit longer.

This can be avoided by modifying your configuration file or how you invoke TopoStats.

In a configuration file is the parameter `base_dir` which if not using your own configuration file defaults to `./` (the
current working directory). The next parameter is `output_dir` which defaults to `./output`

You can modify either of these if for example you have all the images you want to process in the `img/` directory and
don't want to include those that have already been processed (copies of which will reside under `./output`) by
setting...

``` yaml
base_dir: img/
```

When you run TopoStats a second time (e.g. with different parameters) it won't then pick up the images under
`outputs/`. It is also possible to modify `output_dir: ./output_round2/` to ensure you don't over-write the first set of output.

You can also modify this behaviour using the command line options. `topostats --help` shows the following...

``` shell
usage: topostats [-h] [-v] [-c CONFIG_FILE] [-s SUMMARY_CONFIG] [--matplotlibrc MATPLOTLIBRC] [-b BASE_DIR] [-o OUTPUT_DIR] [-l LOG_LEVEL] [-j CORES] [-f FILE_EXT] [--output-stats OUTPUT_STATS] [--channel CHANNEL] [--extract EXTRACT] [--image-set IMAGE_SET [IMAGE_SET ...]]
                 {process,load,filter,grains,grainstats,disordered-tracing,nodestats,ordered-tracing,splining,summary,create-config,bruker-rename} ...

Run various programs relating to AFM data. Add the name of the program you wish to run.

options:
  -h, --help            show this help message and exit
  -v, --version         Report the current version of TopoStats that is installed
  -c CONFIG_FILE, --config-file CONFIG_FILE
                        Path to a YAML configuration file.
  -s SUMMARY_CONFIG, --summary-config SUMMARY_CONFIG
                        Path to a YAML configuration file for summary plots and statistics.
  --matplotlibrc MATPLOTLIBRC
                        Path to a matplotlibrc file.
  -b BASE_DIR, --base-dir BASE_DIR
                        Base directory to scan for images.
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        Output directory to write results to.
  -l LOG_LEVEL, --log-level LOG_LEVEL
                        Logging level to use, default is 'info' for verbose output use 'debug'.
  -j CORES, --cores CORES
                        Number of CPU cores to use when processing.
  -f FILE_EXT, --file-ext FILE_EXT
                        File extension to scan for.
  --output-stats OUTPUT_STATS
                        'basic' (image and grain) or 'full' (image, grain, branch and molecule) statistics written to CSV.
  --channel CHANNEL     Channel to extract.
  --extract EXTRACT     Array to extract when loading '.topostats' files.
  --image-set IMAGE_SET [IMAGE_SET ...]
                        Image set to generate, default is 'core' other options are 'all', 'filters', 'grains', 'grain_crops', 'disordered_tracing', 'nodestats', 'ordered_tracing', 'splining'.

program:
  Available programs, listed below:

  {process,load,filter,grains,grainstats,disordered-tracing,nodestats,ordered-tracing,splining,summary,create-config,bruker-rename}
    process             Process AFM images. Additional arguments over-ride defaults or those in the configuration file.
    load                Load and save all images as .topostats files for subsequent processing.
    filter              Load and filter images, saving as .topostats files for subsequent processing.
    grains              Load filtered images from '.topostats' files and detect grains.
    grainstats          Load images with grains from '.topostats' files and calculate statistics.
    disordered-tracing  WIP DO NOT USE - Skeletonise and prune objects to disordered traces.
    nodestats           WIP DO NOT USE - Calculate node statistics and disentangle molecules.
    ordered-tracing     WIP DO NOT USE - Ordered traces of pruned skeletons.
    splining            WIP DO NOT USE - Splining of traced molecules to produce smooth curves.
    summary             Plotting and summary of TopoStats output statistics.
    create-config       Create a configuration file using the defaults.
    bruker-rename       Append the suffix '.spm' to old (numeric) Bruker file extensions.
```

...and one of the options to the main `topostats` programme is the `-b / --base-dir` option which will (should!) override whatever
is in the default configuration or user specified configuration file. Thus, sticking with the example of the images to
be processed being in the `img/` directory you should be able to run...

``` shell
topostats --base-dir img/ process
```

...and the images that have already been processed once, copies of which now reside in the `output/` directory, won't be
included.

You can also here opt to direct output to a separate directory by using the `-o / --output-dir` flag, e.g.

``` shell
topostats --base-dir img/ --output-dir output_round2 process
```


I hope that helps @phooda1
